bot/exts/utilities/epoch.py

```diff
@@ -0,0 +1,134 @@
+from typing import Optional, Union
+
+import arrow
+import discord
+from dateutil import parser
+from discord.ext import commands
+
+from bot.bot import Bot
+from bot.exts.core.extensions import invoke_help_command
+
+# https://discord.com/developers/docs/reference#message-formatting-timestamp-styles
+STYLES = {"Epoch": ("",),
+          "Short Time": ("t", "h:mm A",),
+          "Long Time": ("T", "h:mm:ss A"),
+          "Short Date": ("d", "MM/DD/YYYY"),
+          "Long Date": ("D", "MMMM D, YYYY"),
+          "Short Date/Time": ("f", "MMMM D, YYYY h:mm A"),
+          "Long Date/Time": ("F", "dddd, MMMM D, YYYY h:mm A"),
+          "Relative Time": ("R",)
+          }
+DROPDOWN_TIMEOUT = 60
+
+
+class DateString(commands.Converter):
+    """Convert a relative or absolute date/time string to an arrow.Arrow object."""
+
+    async def convert(self, ctx: commands.Context, argument: str) -> Union[arrow.Arrow, Optional[tuple]]:
+        """
+        Convert a relative or absolute date/time string to an arrow.Arrow object.
+
+        Try to interpret the date string as a relative time. If conversion fails, try to interpret it as an absolute
+        time. Tokens that are not recognised are returned along with the part of the string that was successfully
+        converted to an arrow object. If the date string cannot be parsed, BadArgument is raised.
+        """
+        try:
+            return arrow.utcnow().dehumanize(argument)
+        except ValueError:
+            try:
+                dt, ignored_tokens = parser.parse(argument, fuzzy_with_tokens=True)
+                return arrow.get(dt), ignored_tokens
+            except parser.ParserError:
+                raise commands.BadArgument(f"`{argument}` Could not be parsed to a relative or absolute date")
+
+
+class Epoch(commands.Cog):
+    """Convert an entered time and date to a unix timestamp."""
+
+    @commands.command(name="epoch")
+    async def epoch(self, ctx: commands.Context, *, date_time: DateString = None) -> None:
+        """
+        Convert an entered date/time string to the equivalent epoch.
+
+        **Relative time**
+            accepted units: "seconds", "minutes", "hours", "days", "weeks", "months", "years"
+            Eg ".epoch in a month 4 days and 2 hours"
+
+        **Absolute time**
+            eg ".epoch 2022/6/15 16:43 -04:00"
+            Absolute times must be entered in descending orders of magnitude.
+            Timezones may take the following forms:
+                Z (UTC)
+                ±HH:MM
+                ±HHMM
+                ±HH
+
+        Times in the dropdown are shown in UTC
+        """
+        if not date_time:
+            await invoke_help_command(ctx)
+            return
+
+        if isinstance(date_time, tuple):
+            ignored_tokens, date_time = list(a.strip() for a in date_time[1] if a.strip()), date_time[0]
```

Comment:
I'm not sure what you mean here. Leave it as a tuple as opposed to converting to a list? We're working with a generator here so it has to be converted. Note that this is outdated code; I've since switched to filter+map. But even then the map object needs to be converted to a list

Reply:
`ignored_tokens` is only used on `"".join()` which works with a generator, so we can remove the list call and keep it as
```py
            ignored_tokens, date_time = (a.strip() for a in date_time[1] if a.strip()), date_time[0]
```

Reply:
I would rather keep it as a list. With a generator, the condition `if ignored_tokens:` is True, even when empty